oss_src/unity/python/sframe/data_structures/sarray_builder.py

```diff
@@ -0,0 +1,147 @@
+from .sarray import SArray
+
+class SArrayBuilder(object):
+    """
+    An interface to incrementally build an SArray element by element.
+
+    Once closed, the SArray cannot be "reopened" using this interface.
+
+    Parameters
+    ----------
+    num_segments : int, optional
+        Number of segments that can be written in parallel.
+
+    history_size : int, optional
+        The number of elements to be cached as history. Caches the last
+        `history_size` elements added with `append` or `append_multiple`.
+
+    dtype : type, optional
+        The type the resulting SArray will be. If None, the resulting SArray
+        will take on the type of the first non-None value it receives.
+        
+    Returns
+    -------
+    out : SArrayBuilder
+
+    Examples
+    --------
+    >>> from graphlab import SArrayBuilder
+
+    >>> sb = SArrayBuilder()
+
+    >>> sb.append(1)
+
+    >>> sb.append([2,3])
+
+    >>> sb.close()
+    dtype: int
+    Rows: 3
+    [1, 2, 3]
+
+    """
+    def __init__(self, num_segments=1, history_size=10, dtype=None):
+        self._builder = UnitySArrayBuilderProxy(glconnect.get_client())
+        if dtype is None:
+            dtype = type(None)
+        self._builder.init(num_segments, history_size, dtype)
+        self._block_size = 1024
+
+    def append(self, data, segment=0):
+        """
+        Append a single element to an SArray.
+
+        Throws a RuntimeError if the type of `data` is incompatible with
+        the type of the SArray. 
+
+        Parameters
+        ----------
+        data  : any SArray-supported type
+            A data element to add to the SArray.
+
+        segment : int
+            The segment to write this element. Each segment is numbered
+            sequentially, starting with 0. Any value in segment 1 will be after
+            any value in segment 0, and the order of elements in each segment is
+            preserved as they are added.
+        """
+        self._builder.append(data, segment)
+        
+    def append_multiple(self, data, segment=0):
+        """
+        Append multiple elements to an SArray.
+
+        Throws a RuntimeError if the type of `data` is incompatible with
+        the type of the SArray. 
+
+        Parameters
+        ----------
+        data  : any SArray-supported type
+            A data element to add to the SArray.
+
+        segment : int
+            The segment to write this element. Each segment is numbered
+            sequentially, starting with 0. Any value in segment 1 will be after
+            any value in segment 0, and the order of elements in each segment is
+            preserved as they are added.
+        """
+        if not hasattr(data, '__iter__'):
+            raise TypeError("append_multiple must be passed an iterable object")
+        tmp_list = []
+        block_pos = 0
+        first = True
+        while block_pos == self._block_size or first:
```

Comment:
It is awkward code. Your way is cleaner. For whatever reason, at the time it felt like the best way. I can change it.

It does work though. Once the block is full, the code breaks from the inside loop.